app/src/main/java/com/example/drawingactivity/DrawActivity.kt

```diff
@@ -89,88 +90,117 @@ class DrawActivity : AppCompatActivity() {
                     onEditText = { layerName -> viewModel.setCurrentText(layerName) },
                     onSubmit = { layerName ->
                         if (layerName != null) {
-                            viewModel.addLayer(layerName)
-                        } else viewModel.dialogCompleted()
+                            viewModel.addToList(layerName)
+                        } else {
+                            viewModel.dialogCompleted()
+                        }
                     },
                 )
-            } else if (!it.isDialogOpen) {
-                isAddDialogOpen = false
-            }
-            if (it.isLayerSheetOpen && !isLayerListViewSheetOpen) {
+            } else isAddDialogOpen = false
+            if (it.isLayerListViewSheetOpen && !isLayerListViewSheetOpen) {
                 isLayerListViewSheetOpen = true
-                viewModel.setLayerListViewState()
-                bottomSheetLambda = dialogUtils.layerListBottomSheet(context = this,
+                onLayerListUpdate = dialogUtils.layerListBottomSheet(context = this,
                     layerList = it.layerList,
-                    onDialogComplete = { viewModel.dialogCompleted() },
+                    onSheetComplete = {
+                        isLayerListViewSheetOpen = false
+                        viewModel.dialogCompleted()
+                    },
                     onDelete = { index ->
-                        viewModel.deleteLayer(index)
-                        if (!viewModel.isLayerListEmpty()) {
-                            bottomSheetLambda?.invoke(index, null, DialogOperation.DELETE)
+                        viewModel.setCurrentIndex(index)
+                        viewModel.deleteDialogOpened()
```

Comment:
Again, this is bleeding view and business logic. The view should be less dumb

Reply:
View now is aware of the model it is building. Overall, the view has more control over what is going on in it and is communicating changes to both the viewModel and the adapter list, who are updating their states based on what the user makes in the view.

I think this should address this issue, if not, I will readjust again